walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,58 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.stream.Collectors;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.stream.Stream;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> possibleTimes = new ArrayList();
+    Stream.concat(events.stream()
+    	.filter(event -> areAnyAttendeeBusy(event, request))
+    	.map(Event::getWhen), Stream.of(TimeRange.fromStartDuration(TimeRange.END_OF_DAY+1, 0)))
+    .sorted(TimeRange.ORDER_BY_START)
+    .reduce(TimeRange.fromStartDuration(0, 0), (previous, current) -> 
+      mergeTimes(previous, current, request, possibleTimes)
+    );
+
+    return possibleTimes;
+  }
+
+  public Boolean areAnyAttendeeBusy(Event event, MeetingRequest request) {
+    return !event.getAttendees()
+      .stream()
+      .filter(attendee->request.getAttendees().contains(attendee))
```

Comment:
Try to avoid explicit calls to `collect` when you can, as this will materialize the stream into an object.

Compare this to a for loop that returns immediately when it finds a busy attendee. This stream approach won't break, and will need space to store all the busy attendees for a given event.

I'd suggest looking at `Stream.anyMatch` which captures the abstraction of "do any elements satisfy a certain property?". This allows for short-circuiting and becomes just as efficient as the imperative version.